[FIX] records: use temp table in replacing refs

jjmaksoud opened the pull request:
For the jsonb company dependent indirect references when replacing record refs, building the query using the original id mapping input can generate a large query based on the number of references being updated, leading to a `stack depth limit exceeded` error.

We can use the mapping saved in the temp table `_upgrade_rrr` instead.

TBG-1752

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/upgrade-util/pull/224.png)](https://mergebot.odoo.com/odoo/upgrade-util/pull/224)

jjmaksoud commented:
upgradeci retry with always only documents_project_sale in version 18.0

aj-fuentes commented:
What examples do we have of such issue? Isn't this a problem of the caller to replace_records_reference_batch? Or that we should implement something perhaps splitting the id mapping?

aj-fuentes commented on src/util/records.py:
```diff
@@ -1544,30 +1544,21 @@ def replace_record_references_batch(cr, id_mapping, model_src, model_dst=None, r
+                    """
+                     UPDATE {table} t
+                        SET {column} = (
+                                SELECT jsonb_object_agg(key, COALESCE((SELECT new FROM _upgrade_rrr WHERE old = value::int), value::int))
```
I'm not sure a subquery here wouldn't end up making this painfully slow.

jjmaksoud commented:
i saw this so far only in documents upgrade when replacing folders with documents [here](https://github.com/odoo/upgrade/blob/6a3075e4c9d533a3e6a6fe15ec6d194eea2f7e82/migrations/documents/saas~17.5.1.4/pre-migrate.py#L193)
example request 2592340: has over 360,000 folder mappings and the new query was fast ( <2s on the test server)
I think the subquery should be okay because the number of companies is expected to be low, the explode bucket size is manageable, and the mapping table has an index

aj-fuentes commented on src/util/records.py:
```diff
@@ -1544,30 +1544,21 @@ def replace_record_references_batch(cr, id_mapping, model_src, model_dst=None, r
+                                FROM jsonb_each_text(t.{column}))
+                    FROM _upgrade_rrr map
+                    WHERE t.{column} IS NOT NULL
+                        AND (SELECT * FROM JSONB_PATH_QUERY(t.{column}, '$.*'))::int = map.old
```
I don't get how this work? Did you mean `map.old IN ...` Don't we get the error of casting jsonb null type to integer?

aj-fuentes replied:
I think this is also updating the same row multiple times. I'd suggest we collect the info to update in a CTE to avoid that.

aj-fuentes replied:
quick test:
```
> select 1 where (SELECT jsonb_path_query FROM JSONB_PATH_QUERY('{"1":3, "2":4, "3":null}'::jsonb, '$.*'))::int = 1
more than one row returned by a subquery used as an expression
```

jjmaksoud replied:
seems the data i tried had a single company per record and no nulls, updated the syntax and tested
I also added values that will be overwritten if updated multiple times but they seem to be correct, checking more into that

KangOl commented:
Can you add a test?

KangOl commented on src/util/records.py:
```diff
@@ -1544,30 +1544,21 @@ def replace_record_references_batch(cr, id_mapping, model_src, model_dst=None, r
+                query = format_query(
+                    cr,
+                    """
+                     UPDATE {table} t
+                        SET {column} = (
+                                SELECT jsonb_object_agg(key, COALESCE((SELECT new FROM _upgrade_rrr WHERE old = value::int), value::int))
+                                FROM jsonb_each_text(t.{column}))
+                    FROM _upgrade_rrr map
+                    WHERE t.{column} IS NOT NULL
+                        AND map.old IN (SELECT JSONB_PATH_QUERY::int FROM JSONB_PATH_QUERY(t.{column}, '$.* ?(@ != null)'))
```
I also don't like those subqueries.
Here is a version using a CTE
```suggestion
                     WITH _upg_cd AS (
                         SELECT t.id, 
                                jsonb_object_agg(j.key, COALESCE(r.new, j.value::int)) as value
                           FROM {table} t
                           JOIN jsonb_each_text(t.{column}) j
                             ON true
                      LEFT JOIN _upgrade_rrr r 
                             ON r.old = j.value::integer
                          WHERE {{parallel_filter}}
                       GROUP BY t.id
                         HAVING bool_or(r.new IS NOT NULL)
                     )
                     UPDATE {table} t
                        SET {column} = u.value
                       FROM _upg_cd u
                      WHERE u.id = t.id
```

KangOl replied:
I don't know how PG will behave with the `LEFT JOIN` after 2 tables in the `FROM`. At least, with a `JOIN ON true`, the behavior is clear.

jjmaksoud commented:
The test depends on https://github.com/odoo/odoo/pull/200509

jjmaksoud commented:
the community PR is merged and the test is now green

KangOl commented on src/base/tests/test_util.py:
```diff
@@ -1364,6 +1364,49 @@ def test_replace_record_references_batch__uniqueness(self):
+        c3 = self.env["res.currency"].create({"name": "RC3", "symbol": "RC3"})
+
+        p1 = self.env["res.partner"].create({"name": "Captain Jack"})
+        p2 = self.env["res.partner"].create({"name": "River Song"})
```
To ensure the `replace_record_references_batch` don't do an UPDATE on all records, I would create a 3rd partner with a `x_test_curr` that shouldn't be updated (`{"4": c4.id}`) and then add a trigger on update of this row.

jjmaksoud replied:
added the trigger, let me know if it's an acceptable testing utility

KangOl commented on src/base/tests/test_util.py:
```diff
@@ -1364,6 +1369,59 @@ def test_replace_record_references_batch__uniqueness(self):
+            c2.id: c3.id,
+            c3.id: c1.id,
+        }
+        add_test_trigger(self.env.cr, "replace_record_references", "res_partner", "UPDATE", f"new.id = {p3.id}")
+        util.replace_record_references_batch(self.env.cr, mapping, "res.currency")
```
This could be context manager
```suggestion
        with self.assertNotUpdated("res_partner", [p3.id]):
            util.replace_record_references_batch(self.env.cr, mapping, "res.currency")
```

Signature: `def assertNotUpdated(self, table, ids=None):`.
When ids is None, the whole should be handled, which mean no `INSERT`.

We can also add the positive version `def assertUpdated(self, table, ids=None)`

These functions need to be added in a dedicated commit.

KangOl replied:
To be consistent with the other `assertX` functions, the signature needs to take a `msg` argument.

`assertNotUpdated(self, table, ids=None, msg=None)`

KangOl commented on src/testing.py:
```diff
@@ -122,6 +125,89 @@ def setUpClass(cls):
+                FOR EACH ROW {cond} EXECUTE
+           FUNCTION fail_assert_not_updated()
+                """,
+                when=sql.SQL("UPDATE" if ids else "UPDATE or INSERT"),
```
```suggestion
                when=sql.SQL("UPDATE" if ids is not None else "UPDATE or INSERT"),
```

KangOl commented on src/testing.py:
```diff
@@ -122,6 +125,89 @@ def setUpClass(cls):
+                """,
+                when=sql.SQL("UPDATE" if ids else "UPDATE or INSERT"),
+                table=table,
+                cond=sql.SQL("WHEN (new.id IN %s)" if ids else ""),
```
```suggestion
                cond=sql.SQL("WHEN (new.id = ANY(%s))" if ids else ""),
```

KangOl commented on src/testing.py:
```diff
@@ -122,6 +125,89 @@ def setUpClass(cls):
+                table=table,
+                cond=sql.SQL("WHEN (new.id IN %s)" if ids else ""),
+            ),
+            (tuple(ids) if ids else None,),
```
```suggestion
            [list(ids) if ids is not None else None],
```

KangOl commented on src/testing.py:
```diff
@@ -122,6 +125,89 @@ def setUpClass(cls):
+                FOR EACH ROW {cond} EXECUTE
+           FUNCTION save_updated()
+                """,
+                when=sql.SQL("UPDATE" if ids else "UPDATE or INSERT"),
+                table=table,
+                cond=sql.SQL("WHEN (NEW.id IN %s)" if ids else ""),
+            ),
+            (tuple(ids) if ids else None,),
```
Ditto.

KangOl commented on src/testing.py:
```diff
@@ -122,6 +125,89 @@ def setUpClass(cls):
+            yield
+            cr.execute("SELECT id FROM _upg_test_upd_id")
+            updated_ids = [r[0] for r in cr.fetchall()]
+            if not ids:
+                self.assertNotEqual(updated_ids, [], "No records were updated")
+            else:
+                for id in ids:
+                    self.assertIn(id, updated_ids, f"Record with id={id} was not updated")
```
Raise directly `AssertionError` instead of relying on `self.assert*` functions.

KangOl commented on src/testing.py:
```diff
@@ -122,6 +125,89 @@ def setUpClass(cls):
+            ),
+            (tuple(ids) if ids else None,),
+        )
+        try:
+            yield
+        finally:
```
Not sure the right exception is raised.

KangOl reviewed (changes requested):
Maybe add tests for the new `assert` function to expect they behave correctly in all expected cases.

KangOl commented on src/base/tests/test_util.py:
```diff
@@ -1364,6 +1364,58 @@ def test_replace_record_references_batch__uniqueness(self):
+            p3.id: f'{{"1":{c4.id}}}',
+        }
+        for id, value in old.items():
+            self.env.cr.execute(
+                """
+                UPDATE res_partner
+                  SET x_test_curr = %s
+                WHERE id = %s
+                """,
+                [value, id],
+            )
```
Nit. Can be written on one line.
```suggestion
            self.env.cr.execute("UPDATE res_partner SET x_test_curr = %s WHERE id = %s", [value, id])
```

jjmaksoud commented:
Updated the assert methods, and ended up using a temp table for the assertNotUpdated as well instead of raising a psql exception.
Also added some test cases for the new assert functions. They can be used together, but the known limitation at the moment is that the same method cannot be used at the same time for a multiple tables.

KangOl reviewed (approved):
@robodoo rebase-ff r+

![](https://cdn2.thecatapi.com/images/des.jpg)

robodoo commented:
Merge method set to rebase and fast-forward.

robodoo commented:
@jjmaksoud @KangOl staging failed: ci/l10n on 848319d8abe9ae140bc7657ae39bfc1b1b8fbd99 (view more at https://runbot.odoo.com/runbot/build/79745018)

d-fence commented:
robodoo retry
(failing test neutralized)
